mesonbuild/modules/unstable_icestorm.py

```diff
@@ -12,78 +12,119 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from __future__ import annotations
+import itertools
+import typing as T
+
+from . import ExtensionModule, ModuleReturnValue
+from .. import build
 from .. import mesonlib
-from ..interpreterbase import flatten
+from ..interpreter.type_checking import CT_INPUT_KW
 from ..interpreterbase import FeatureNew
+from ..interpreterbase.decorators import KwargInfo, typed_kwargs, typed_pos_args
+
+if T.TYPE_CHECKING:
+    from typing_extensions import TypedDict
+
+    from . import ModuleState
+    from ..interpreter import Interpreter
+    from ..programs import ExternalProgram
+
+    class ProjectKwargs(TypedDict):
 
-from . import ExtensionModule
+        sources: T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        constraint_file: T.Union[mesonlib.FileOrString, build.GeneratedTypes]
 
 class IceStormModule(ExtensionModule):
 
     @FeatureNew('FPGA/Icestorm Module', '0.45.0')
-    def __init__(self, interpreter):
+    def __init__(self, interpreter: Interpreter) -> None:
         super().__init__(interpreter)
-        self.yosys_bin = None
+        self.yosys_bin: T.Optional[ExternalProgram] = None
+        self.arachne_bin: T.Optional[ExternalProgram] = None
+        self.icepack_bin: T.Optional[ExternalProgram] = None
+        self.iceprog_bin: T.Optional[ExternalProgram] = None
+        self.icetime: T.Optional[ExternalProgram] = None
```

Comment:
In other modules we turned this into a `self.tools: T.Dict`.

Reply:
Yeah, that seems better.